src/Security/InheritedPermissions.php

```diff
@@ -0,0 +1,632 @@
+<?php
+
+namespace SilverStripe\Security;
+
+use InvalidArgumentException;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\Hierarchy\Hierarchy;
+use SilverStripe\Versioned\Versioned;
+
+/**
+ * Calculates batch permissions for nested objects for:
+ *  - canView: Supports 'Anyone' type
+ *  - canEdit
+ *  - canDelete: Includes special logic for ensuring parent objects can only be deleted if their children can
+ *    be deleted also.
+ */
+class InheritedPermissions
```

Comment:
I think that this would make more sense if it implemented a more generic interface such as `PermissionChecker`, `AccessChecker`, `PermissionStore`, etc.

The interface that I would recommend is probably:

 * checkOne($permission, $id, $member);
 * checkMany($permission, $ids, $member);

That means that an inherited permission model is simply 1 approach that we're plugging into this; you could provide other implementations of the same interface.

It will be a good starting point for ACL work in the future, without requiring too much more work in the short term.